src/MobxPromise.ts

```diff
@@ -153,38 +152,52 @@ export class MobxPromiseImpl<R>
 				if (error)
 					return error;
 
-		return this.internalError;
+		return this.internalError.value;
 	}
 
 	/**
 	 * This lets mobx determine when to call this.invoke(),
 	 * taking advantage of caching based on observable property access tracking.
 	 */
-	@computed private get latestInvokeId()
+	@computed private get invokeResult()
 	{
-		window.clearTimeout(this._latestInvokeId);
-		let promise = this.invoke();
-		let invokeId:number = window.setTimeout(() => this.setPending(invokeId, promise));
-		return this._latestInvokeId = invokeId;
+		let syncResult = {
+			isSet: false,
+			value: (undefined as R|undefined)
+		};
+
+		let promise = this.invoke((result:R)=>{
+			syncResult.isSet = true;
+			syncResult.value = result;
+		});
+
+		// override promise value, and set synchronously, if theres a synchronous resolution
+		if (syncResult.isSet) {
+			this.setComplete(this._asyncInvokeCount, syncResult.value!);
+			return 0;
+		} else {
+			this._asyncInvokeCount += 1;
+			this.setPending(this._asyncInvokeCount, promise);
+			return 0;
+		}
 	}
 
 	@action private setPending(invokeId:number, promise:PromiseLike<R>)
 	{
-		this.invokeId = invokeId;
 		promise.then(
 			result => this.setComplete(invokeId, result),
 			error => this.setError(invokeId, error)
 		);
-		this.internalStatus = 'pending';
+		this.internalStatus.value = 'pending';
 	}
 
 	@action private setComplete(invokeId:number, result:R)
 	{
-		if (invokeId === this.invokeId)
+		if (invokeId === this._asyncInvokeCount)
```

Comment:
deserves a comment to explain that asyncInvokeCount is used to make sure we only use the latest invokation